zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,193 @@
+    File[] dirs = backupDir.listFiles(fileFilter);
+
+    if (dirs == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get directory list.");
+    }
+    return Arrays.asList(dirs);
+  }
+
+  @Override
+  public void copyToBackupStorage(File srcFile, File destName) throws IOException {
+    InputStream inputStream = null;
+    OutputStream outputStream = null;
+    String backupTempFilePath;
+    File backupTempFile;
+
+    if (!srcFile.exists() || srcFile.isDirectory()) {
+      throw new BackupException("File " + srcFile.getPath()
+          + " does not exist or is a directory, cannot be copied to backup storage.");
+    }
+    try {
+      inputStream = new FileInputStream(srcFile);
+
+      backupTempFilePath =
+          constructBackupFilePath(BackupStorageUtil.constructTempFileName(destName.getName()));
+      backupTempFile = new File(backupTempFilePath);
+
+      BackupStorageUtil.createFile(backupTempFile, true);
+      outputStream = new FileOutputStream(backupTempFile);
+
+      BackupStorageUtil.streamData(inputStream, outputStream);
+    } finally {
+      if (inputStream != null) {
+        inputStream.close();
+      }
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+
+    Files
+        .move(Paths.get(backupTempFilePath), Paths.get(constructBackupFilePath(destName.getName())),
+            StandardCopyOption.REPLACE_EXISTING);
+    BackupStorageUtil.cleanUpTempFiles(backupTempFile.getParentFile());
+  }
+
+  @Override
+  public void copyToLocalStorage(File srcName, File destFile) throws IOException {
+    InputStream inputStream = null;
+    OutputStream outputStream = null;
+
+    // Create input stream from the source file in backup storage
+    String backupFilePath = constructBackupFilePath(srcName.getName());
+    File backupFile = new File(backupFilePath);
+
+    if (!backupFile.exists() || backupFile.isDirectory()) {
+      throw new BackupException("The backed up file with the file path " + backupFilePath
+          + " does not exist or is a directory, could not copy to local storage.");
+    }
+    try {
+      inputStream = new FileInputStream(backupFile);
+
+      BackupStorageUtil.createFile(destFile, true);
+      outputStream = new FileOutputStream(destFile);
+
+      BackupStorageUtil.streamData(inputStream, outputStream);
+    } finally {
+      if (inputStream != null) {
+        inputStream.close();
+      }
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+  }
+
+  @Override
+  public void delete(File fileToDelete) throws IOException {
+    String backupFilePath = constructBackupFilePath(fileToDelete.getPath());
+    Files.deleteIfExists(Paths.get(backupFilePath));
+  }
+
+  private String constructBackupFilePath(String fileName) {
+    return BackupStorageUtil.constructBackupFilePath(fileName, fileRootPath);
+  }
```

Comment:
Making this a private method might not be necessary since it's already a one-liner